fix(#887): complete strictMode for JSONArray

rikkarth opened the pull request:
### Description

If in strict mode, and having reached the end of the **JSONArray** construction life-cycle with no errors, we then perform an input validation.
![image](https://github.com/stleary/JSON-java/assets/76500344/2afde4f6-156e-48b6-8df4-18dfadeed637)

In this validation we explicitly check if (1.) the last bracket was found and (2.) if more characters exist after that.

If the next character is not the end of the array (0), then it means we have more characters after the final closing square bracket, so we throw an error message informing the array is not compliant and we give back the complete input to the user for easy debugging.

![image](https://github.com/stleary/JSON-java/assets/76500344/8c9c8ecb-3c15-48f4-9a78-131dafdb042c)

To ensure this logic doesn't affect 2D arrays I also created a new test for it.
![image](https://github.com/stleary/JSON-java/assets/76500344/bb5c52fd-2c77-4d3e-b513-7c98006447e7)

From the examples provided in the issue created: https://github.com/stleary/JSON-java/issues/887

### Additional Note
Since the logic is fully encapsulated inside `validateInput()` which is only run if strictMode is true, it is guaranteed that previous implementations are not affected when strictMode is false and that this implementation is modular and can be maintained, improved or removed by simply removing `validateInput()`.

### Test cases added
![image](https://github.com/stleary/JSON-java/assets/76500344/4b85e096-3e1d-43f8-b562-4cca39df8043)

![image](https://github.com/stleary/JSON-java/assets/76500344/f18dca65-ec16-464e-8df2-19b7f7c9b5dc)

stleary commented on src/main/java/org/json/JSONArray.java:
```diff
@@ -144,6 +144,42 @@ public JSONArray(JSONTokener x, JSONParserConfiguration jsonParserConfiguration)
+        boolean nextCharacterIsNotEoF = x.nextClean() != 0;
+
+        if (isEndOfArray && nextCharacterIsNotEoF) {
+            String completeInput = collectCompleteInput(x);
```
It's probably not a good idea to collect complete input, in the case of a large JSONArray. Just the char value, index, and line number should be sufficient.

rikkarth replied:
Will do.

stleary commented on src/main/java/org/json/JSONArray.java:
```diff
@@ -144,6 +144,42 @@ public JSONArray(JSONTokener x, JSONParserConfiguration jsonParserConfiguration)
         }
+
+        if (jsonParserConfiguration.isStrictMode()) {
+            validateInput(x);
```
Please use similar logic to the ']' case starting on line 133.

stleary commented on src/test/java/org/json/junit/JSONParserConfigurationTest.java:
```diff
@@ -208,6 +218,15 @@ public void verifyMaxDepthThenDuplicateKey() {
+            "[],",
+            "[]:",
+            "[],[",
+            "[],{",
```
Please add this entry to the noncompliant list:
`"[[a]]"`
It will break the unit tests since JSONTokener.nextValue() does not forward the jsonParserConfiguration object when creating a new JSONArray, so some additional work will be needed.

rikkarth replied:
Will do.

stleary commented:
@rikkarth Sorry for taking so long to complete the review. Comments added.

rikkarth commented:
1. Fixed issue with double array breaking JSONTokener.nextValue() which now forwards jsonParserConfiguration object when creating a new JSONArray.
2. Changed input validation in JSONArray, removing collection of complete input in favor of leveraging x.syntaxError().

![image](https://github.com/stleary/JSON-java/assets/76500344/5aaa94ad-6ee8-43fb-bd0f-aa16afa84d18)

![image](https://github.com/stleary/JSON-java/assets/76500344/c6cd0f5d-7530-4bb4-b03f-0e473bb1e478)

stleary commented:
@rikkarth  Please merge from master to pick up the latest changes:
![image](https://github.com/stleary/JSON-java/assets/5107558/2af89525-af3d-484b-8acb-f4fc8c06e754)

stleary commented:
@rikkarth Looks like there is a regression when parsing array elements with non-string simple values. Please add this unit test:
```
    @Test
    public void shouldHandleNumericArray() {
        String expected = "[10]";
        JSONArray jsonArray = new JSONArray(expected, new JSONParserConfiguration().withStrictMode(true));
        assertEquals(expected, jsonArray.toString());
    }
```

This test passes when strict mode is false, but fails with this output when strict mode is true:
![image](https://github.com/stleary/JSON-java/assets/5107558/ba073fed-ad8d-4626-870f-8c16c5d6ccc9)

Similar problems occur in strict mode when the array elements are true/false, and null.

rikkarth commented:
Also boolean values are producing the same behaviour, I will fix it.
![image](https://github.com/stleary/JSON-java/assets/76500344/e77525b7-f88c-443b-8af4-0bd99eac141b)

rikkarth commented:
I changed how I validated numbers, boolean and string values using existing logic, I've also added what I believe a few, more robust, test cases to prevent more regressions.

stleary commented:
null without quotes is a valid JSON value, just like true and false (see https://json.org). Strict mode should allow this value when it appears without quotes. Please add this test case for coverage:
```
    @Test
    public void allowNullInStrictMode() {
        String expected = "[null]";
        JSONArray jsonArray = new JSONArray(expected, new JSONParserConfiguration().withStrictMode(true));
        assertEquals(expected, jsonArray.toString());
    }

```

rikkarth commented:
Done as requested. Now it was simple as adding it to the this method:

```java
    private Object getValidNumberOrBooleanFromObject(Object value) {
        if (value instanceof Number || value instanceof Boolean || value.equals(JSONObject.NULL)) {
            return value;
        }

        throw new JSONException(String.format("Value is not surrounded by quotes: %s", value));
    }
```

stleary commented:
@rikkarth, This string fails in strict mode because there are chars after the end of an empty array:
`String s = "[[]]";`
This string is allowed in strict mode because the chars at the end are not validated:
 `String s = "[1],";`
You might be able to address the first problem by making the parser code block starting at line 160 exactly the same as the code block starting at line 134, as I mentioned sometime back. But it will probably still need to be reworked to handle values after the closing bracket, like JSONObject. The second problem means that the parser is still not reliably detecting invalid trailing chars. If the current implementation is unable to do that, it might be necessary to revisit the design for this feature.

rikkarth commented:
> @rikkarth, This string fails in strict mode because there are chars after the end of an empty array: `String s = "[[]]";` This string is allowed in strict mode because the chars at the end are not validated: `String s = "[1],";` You might be able to address the first problem by making the parser code block starting at line 160 exactly the same as the code block starting at line 134, as I mentioned sometime back. But it will probably still need to be reworked to handle values after the closing bracket, like a closing JSONObject brace. The second problem means that the parser is still not reliably detecting invalid trailing chars. If the current implementation is unable to do that, it might be necessary to revisit the design for this feature.

Yes you are absolutely right. I will come back to you with a more solid, robust solution and design.

Thank you for your patience.

rikkarth commented:
I'm near a solution for this I will share my results and implementation soon.

rikkarth commented:
Mr @stleary, 

After a few days of serious research of the codebase and work I believe I came up to a nice solution.

Ideally I'd like to present you with the changes I performed over a call, but I'll try to do my best through here.

Will present it in the next comment.

rikkarth commented:
# * DRAFT *

stleary commented:
@rikkarth Please see #891. Let's pause this review until the unit tests have been updated to give a better view of strict mode behavior.

stleary commented:
Branch [strict-mode-unit-tests](https://github.com/stleary/JSON-java/tree/strict-mode-unit-tests) contains all of your code from this branch. In the JSONParserConfiguration default constructor, I forced strictMode to be true, and then modified the broken unit tests or replaced @Test with @Ignore. 
* Most of the broken tests were due to inadvertently forgetting to use double quotes in otherwise valid JSON docs. You can ignore those changes. 
* A few tests were set to @Ignore because they intentionally used unquoted text. They will be restored later, and are not a concern. 
* There were a few grey areas, e.g. when an invalid numeric value is supposed to be converted to string, but instead throws an exception. It could be argued either way - should strictMode allow the conversion or not? No decision has been made yet, no action needed at this time. 
* There is a regression in single quote handling within double-quote delimited strings that must be fixed in strictMode.
* There are also several errors in JSONTokenerTest that suggest a regression.

You can fix these issues on your current branch, or you can take a new fork off of the strict-mode-unit-tests branch.

rikkarth commented:
I will review all this points and will come back to you with. We feedback.

Thank you for this analysis.

rikkarth commented:
@stleary 

**Update**:
I fixed all the test cases in JSONTokener apart from the last which I require more feedback:

![image](https://github.com/stleary/JSON-java/assets/76500344/97de6c12-11e2-405a-a72d-89693793b11e)

## The Bug

The issue was, I didn't perform the most simple of validations which checks for end of file after finding ']' at 2nd, which I included now. After doing it, all tests pass accordingly.
![image](https://github.com/stleary/JSON-java/assets/76500344/24d8d05b-c1f4-46af-b4b4-47029f74d9b7)

As for this test case

```java
@Test
public void testValid() {
    // ...
    checkValid("1 2", String.class);
}

throws `JSONException` which is indeed valid, since you manually set `strictMode` true in `JSONParserConfiguration`.

The test case value is not inside an array;
The test case value is not inside quotes;

What should I do with this case?
```
org.json.JSONException: Value '1 2' is not surrounded by quotes at 3 [character 4 line 1]
```

rikkarth commented:
This test is failing because the expected testCase is actually malformed, strictMode working accordingly
![image](https://github.com/stleary/JSON-java/assets/76500344/6bd530b9-87d4-4985-9084-9323c1bb386e)

![image](https://github.com/stleary/JSON-java/assets/76500344/3b6cbd2c-ab4d-4d8c-9a74-2ef5b695552b)

rikkarth commented:
> Branch [strict-mode-unit-tests](https://github.com/stleary/JSON-java/tree/strict-mode-unit-tests) contains all of your code from this branch. In the JSONParserConfiguration default constructor, I forced strictMode to be true, and then modified the broken unit tests or replaced @test with @ignore.

All of the points you've highlighted were fixed. There are a few cases (e.g `checkValid("1 2", String.class);`) where I don't exactly know what I should do there, and will require some more feedback on this.

I removed the TODO's, comments, etc that you provided in [strict-mode-unit-tests](https://github.com/stleary/JSON-java/tree/strict-mode-unit-tests) as I fixed each issue.

I left some of the "important" TODO's the one's you wrote "TBD later" or "will revert later" so you could easily navigate back to them later.

> There is a regression in single quote handling within double-quote delimited strings that must be fixed in strictMode.

Fixed.

> There are also several errors in JSONTokenerTest that suggest a regression.

Fixed.

> You can fix these issues on your current branch, or you can take a new fork off of the strict-mode-unit-tests branch.

I merged them into this branch to keep a linear history of what was affected in this PR.

rikkarth commented:
I added this TODO's because I strongly believe the test cases here are super bloated and could be hugely simplified.

I will make my case in a future PR, leaving this here for reference.
![image](https://github.com/stleary/JSON-java/assets/76500344/7b334b41-652d-444e-aca7-44924c9fa1f1)

stleary commented:
@rikkarth Nice work. 
The `1 2` test result signals a regression - in this case, a change in the behavior of a public API, which is always a concern. Might be good to see if the existing behavior can be retained.
Will review the code, and add some comments.

rikkarth commented:
> @rikkarth Nice work. The `1 2` test result signals a regression - in this case, a change in the behavior of a public API, which is always a concern. Might be good to see if the existing behavior can be retained. Will review the code, and add some comments.

The test for this use-case passes in `strictMode=false` just not in `strictMode=true` which I think makes sense under JSON schema.
![image](https://github.com/stleary/JSON-java/assets/76500344/23366cbb-4051-4d1a-9294-1cd28a03c0da)

The tokener is trying to understand if the value is enclosed within an array or object.
![image](https://github.com/stleary/JSON-java/assets/76500344/05234cf2-a539-4cd6-8295-24c935491616)

This is the return when `strictMode=false` when:
```java
new JSONArray("1 2");
```
![image](https://github.com/stleary/JSON-java/assets/76500344/5aabe979-a5f1-4fc8-84c6-ba603858bd8c)

rikkarth commented:
Could the label for this PR be updated? It's no longer pending redesign but pending review.

Thank you.

stleary commented:
> Could the label for this PR be updated? It's no longer pending redesign but pending review.

I think the label still applies. The goal of issue #887 was just to catch trailing chars at the end of the JSON doc. There might be a simpler solution that hasn't been found yet. For example, adding properties to the tokener does not seem like the right direction, or the way that JSONArray() was rewritten.

stleary commented:
> Did you have time to read my previous comment?
> It is the last point I have to complete the points you've defined previously.
> I can make this last test pass, I'm just wondering why should this test pass in strict mode?
> Do you have an opinion on it?

Do you mean the "1 2" validity test? Do you think the result difference is due to code changes to enforce quote-delimited strings, or to detecting trailing chars? In the former case, no additional changes are needed. In the latter case, it should be handled by the redesign. In either case, you don't need to make it work now.

rikkarth commented:
> > Did you have time to read my previous comment?
> > It is the last point I have to complete the points you've defined previously.
> > I can make this last test pass, I'm just wondering why should this test pass in strict mode?
> > Do you have an opinion on it?
> 
> Do you mean the "1 2" validity test? Do you think the result difference is due to code changes to enforce quote-delimited strings, or to detecting trailing chars? In the former case, no additional changes are needed. In the latter case, it should be handled by the redesign. In either case, you don't need to make it work now.

Yes, the error is exactly that. Quote-delimited strings, which ... makes sense.. If you put quotes around the string it passes the test.

Please check screenshots bellow.

![image](https://github.com/stleary/JSON-java/assets/76500344/d16bfbe2-948e-4edc-ba91-04d5eaf99581)

![image](https://github.com/stleary/JSON-java/assets/76500344/a7cbdf6d-f432-48bc-a58d-b8a297e47562)

stleary commented:
**What problem does this code solve?**
Complete the implementation of strict mode for JSONArray

**Does the code still compile with Java6?**
Yes

**Risks**
Low

**Changes to the API?**
No

**Will this require a new release?**
No

**Should the documentation be updated?**
No

**Does it break the unit tests?**
No, new unit tests were added

**Was any code refactored in this commit?**
No

**Review status**
**APPROVED** 

Starting 3-day comment window

rikkarth commented:
Cleaned up strictMode true flag as instructed.
